pytest_socket.py

```diff
@@ -206,20 +232,29 @@ def resolve_hostnames(hostname: str) -> typing.Set[str]:
 
 def normalize_allowed_hosts(
     allowed_hosts: typing.List[str],
+    resolution_cache: typing.Optional[typing.Dict[str, typing.List[str]]] = None,
 ) -> typing.Dict[str, typing.Set[str]]:
     """Map all items in `allowed_hosts` to IP addresses."""
+    if resolution_cache is None:
+        resolution_cache = {}
     ip_hosts = defaultdict(set)
     for host in allowed_hosts:
         host = host.strip()
         if is_ipaddress(host):
             ip_hosts[host].add(host)
-        else:
-            ip_hosts[host].update(resolve_hostnames(host))
+            continue
+        if host not in resolution_cache:
+            resolution_cache[host] = resolve_hostnames(host)
+        ip_hosts[host].update(resolution_cache[host])
 
     return ip_hosts
 
 
-def socket_allow_hosts(allowed=None, allow_unix_socket=False):
+def socket_allow_hosts(
+    allowed: typing.Union[str, typing.List[str], None] = None,
+    allow_unix_socket: bool = False,
+    resolution_cache: typing.Optional[typing.Dict[str, typing.List[str]]] = None,
```

Comment:
:bug: This is proof that a test is required, because it should be passed down.

Now that you've approved the approach I'll add that! Thank you for the quick turn-around.